plans/main.fmf

```diff
@@ -19,6 +19,7 @@ prepare+:
       #
```

Comment:
No experience with pipx, sorry. Can you show me how would it help?

Reply:
Sure, this is my "workaround" for running pytest: https://gitlab.com/rh-kernel-stqe/sts/-/blob/9be0f8d74ffb1ebc430bae5c0bee43cdb29bf361/plans/main.fmf

Reply:
I forgot the reason, why I'm installing pipx using pip, but 'python3-pipx' should be available in Fedora and EPEL9.

Reply:
Offtopic, but can you run `pipx` on local source as if you've done `pip install -e /path/to/local/source`?

Reply:
Then I'm using `/root/.local/bin/pytest` to avoid issues with PATH and rpm pytest

Reply:
Hm, this will need more thinking... Because, `pytest-container` is recorded in `pyproject.toml`, so, why do I even bother installing it like this? This stinks, I probably took wrong turn somewhere :/

Reply:
> Offtopic, but can you run `pipx` on local source as if you've done `pip install -e /path/to/local/source`?

It also has `--editable` argument

Reply:
@happz I'm looking at it now. As for the existing main branch, I would just do https://gist.github.com/martinhoyer/95de1f435ac3eee0032708dc79b98b16 .

Looking at the PR, I don't see you actually importing pytest_container anywhere, can you point me towards the place it's being used?

Reply:
> Looking at the PR, I don't see you actually importing pytest_container anywhere, can you point me towards the place it's being used?

@martinhoyer expand tests/unit/test_package_managers.py:

```python
import _pytest.logging
import pytest
from pytest_container import Container
from pytest_container.container import ContainerData
```

Then it's used as a fixture, and I think this was the reason why I tried to include it. It needs to be importable even if those tests would be skipped. I plan to look into this tomorrow because adding the requirement here does not feel right.

Reply:
Thanks. Indeed it does not feel right. Regardless of where you'll add the requirement, instead of requiring python3-pytest and others, try require: pipx instead and do:
```
pipx install pytest
pipx inject pytest pytest-container pytest-xdist
```

Reply:
Got it. We're not using `hatch` when running unit tests in CI. It's system Python and system pytest and system packages, installed from RPMs. Which I'm not sure is the right way to do it.

Reply:
I was thinking about doing this, i.e. use `hatch` & let it do its magic:

```diff
-      python3 -m pytest -vvv -ra --showlocals -n 0 -k 'not containers'
+      hatch -vv run test:pytest -vvv -ra --showlocals -n 0 -m 'not containers'
```

Any trap waiting to happen?

Reply:
Can't think of any, as long as hatch is installed. Even epel9 should work afaik

Reply:
Of course there is the matter of pulling unnecessary packages, but I don't know about tmt test plans to say what's needed for what. :)

Reply:
But then how is it tested against the dependency versions on each OS? It would be nice to duplicate the plans or contextualize them so that packit/downstream + testing-farm test against packaged versions and have a separate runner for CI with PyPI and pre-releases.

Reply:
Yep, realized that this morning, that there might be a value in running unit tests against system packages. I might need to use `pipx` for this after all...

Reply:
fwiw, hatch env can also be 'detached' https://hatch.pypa.io/latest/config/environment/overview/#detached-environments. 

This is the venv with system-site-packages way:

```
dnf install python3-pytest
 ... python3-pytest-6.2.2-6.el9.noarch

python3 -m venv ./test --system-site-packages
./test/bin/pip install pytest-container
./test/bin/python -m pytest
./test/bin/python -m pytest

platform linux -- Python 3.9.18, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /root
plugins: testinfra-10.1.0, pytest_container-0.3.0

```

Reply:
Longer term, tmt should provide a `python` framework to control all of these:
- Get the dependencies from optional-dependencies
- Add dependencies to `requires`, download from pypi, add `--pre`, etc.
- Setup a venv
- Run pytest

Right now the `framework` is a test-level argument, but in these cases it would be more helpful to put it on a plan level as well for cases where the framework can be shared across tests.\

For `tmt` it could encounter some dog-fooding issues, but still it would make the intent much clearer

Reply:
Ok, got the tests to pass, finally \o/ Instead of the initial `/tests/unit`, we get the following tests:

```
/tests/unit/with-development-packages/basic
/tests/unit/with-development-packages/extended
/tests/unit/with-system-packages/basic
/tests/unit/with-system-packages/extended
```

I'm pretty sure the `with-development-packages` is fine, it's just a `hatch run`, but I'd appreciate any advice WRT `with-system-packages`. After creating the virtualenv, I had to "symlink" `pytest` into it - yes, it contains system packages, but not scripts installed by these system packages. And I keep "symlinked" in quotes because a mere symlink did not work, as the system `pytest` often includes `#!/usr/bin/python3` shebang which ruins the execute-inside-virtualenv attempt. But it works.

Reply:
Looking...
@happz just to clarify. Is the goal to use rpm pytest as well?

Reply:
I can try it, I didn't go through `python -m pytest` way. Let's see...

Reply:
I'm thinking, since there are non-rpm pytest plugins anyway, it might as well be all PyPI pytest.
In the end, we're testing tmt and it's integration, not pytest itself, no?
 
>  I'm worried that if I tried to install it into the venv with site-packages

Don't think site-packages would be appropriate in this case, if it would just be for xdist plugin.
The pipx example above still stands :) 
```
dnf install pipx
pipx install pytest
pipx inject pytest pytest-xdist pytest-container
```
Note to either uninstall rpm pytest or use full path:

> The default app location is /root/.local/bin and can be
> overridden by setting the environment variable `PIPX_BIN_DIR`.
>

Reply:
`python -m pytest` works \o/ https://artifacts.dev.testing-farm.io/23d6898b-4a98-4a6e-a876-eff22527f027/work-coreboxqx687/plans/features/core/execute/data/guest/default-0/tests/unit/with-system-packages/basic-42/output.txt

Reply:
I'd say this conversation is now resolved - things work, can be improved later.